### PR TITLE
Custom endpoint configuration

### PR DESCRIPTION
**Pull request by kalafut**
### Overview

This change allows the [service endpoints](https://cloud.google.com/apis/design/glossary#api_service_endpoint) used by the underlying client to be customized. This allows specific requests made by the plugin to target alternative service endpoints for use in [Private Google Access](https://cloud.google.com/vpc/docs/configure-private-google-access) environments (+ potential other use cases). The API allows for the service endpoints to be customized at the subdomain level.

The default service endpoints that can be overridden are:

- https://www.googleapis.com
- https://iam.googleapis.com
- https://cloudresourcemanager.googleapis.com
- https://compute.googleapis.com

**Example Usage**
```
$ vault write auth/gcp/config \
    credentials="@vault-auth-tester.json" \
    custom_endpoint=api="https://example-api.com" \
    custom_endpoint=iam="https://example-iam.com" \
    custom_endpoint=crm="https://example-crm.com" \
    custom_endpoint=compute="https://example-compute.com"

$ vault read auth/gcp/config

Key                Value
---                -----
client_email       redacted
client_id          redacted
custom_endpoint    map[api:https://example-api.com compute:https://example-compute.com crm:https://example-crm.com iam:https://example-iam.com]
gce_metadata       [instance_creation_timestamp instance_id instance_name project_id project_number role service_account_id service_account_email zone]
iam_metadata       [project_id role service_account_id service_account_email]
private_key_id     redacted
project_id         redacted
```

### Testing

I've manually tested this change with various custom endpoint configurations to ensure they're used by the correct clients. I'm currently writing additional tests to include with this PR.

### CODE REVIEW

File: plugin/gcp_config.go [hunk 1]
```diff
@@ -21,6 +20,15 @@ type gcpConfig struct {
 	IAMAuthMetadata *authmetadata.Handler   `json:"iam_auth_metadata_handler"`
 	GCEAliasType    string                  `json:"gce_alias"`
 	GCEAuthMetadata *authmetadata.Handler   `json:"gce_auth_metadata_handler"`
+
+	// APICustomEndpoint overrides the service endpoint for www.googleapis.com
+	APICustomEndpoint string `json:"api_custom_endpoint"`
```
Comment (austingebauer) [inline plugin/gcp_config.go:25]: Not sure on name here. There isn't a service I can pinpoint that maps directly to the `www` subdomain. This is used to override the default service endpoints in [go-gcp-common](https://github.com/hashicorp/go-gcp-common/blob/master/gcputil/credentials.go#L31).

File: plugin/gcp_config.go [hunk 2]
```diff
@@ -67,25 +75,40 @@ func (c *gcpConfig) Update(d *framework.FieldData) error {
...
 	}
+
 	if err := c.GCEAuthMetadata.ParseAuthMetadata(d); err != nil {
-		return errwrap.Wrapf("failed to parse gce metadata: {{err}}", err)
+		return fmt.Errorf("failed to parse gce metadata: %w", err)
 	}
+
+	rawEndpoint, exists := d.GetOk("custom_endpoint")
+	if exists {
+		for k, v := range rawEndpoint.(map[string]string) {
```
Comment (austingebauer) [inline plugin/gcp_config.go:96]: Wondering if there is any validation we should do on `v`. Looking closer at requirements for the `option.WithEndpoint()` that's provided as an option to clients.

File: plugin/backend.go [hunk 3]
```diff
@@ -107,16 +113,16 @@ func (b *GcpAuthBackend) IAMClient(s logical.Storage) (*iam.Service, error) {
 // https://pkg.go.dev/google.golang.org/api@v0.45.0/iamcredentials/v1#pkg-overview
 //
 // The client is cached.
-func (b *GcpAuthBackend) IAMCredentialsClient(s logical.Storage) (*iamcredentials.Service, error) {
-	httpClient, err := b.httpClient(s)
+func (b *GcpAuthBackend) IAMCredentialsClient(ctx context.Context, s logical.Storage) (*iamcredentials.Service, error) {
+	opts, err := b.clientOptions(ctx, s, "")
```
Comment (austingebauer) [inline plugin/backend.go:117]: This is only used in test code to sign a JWT. That's why there is no custom endpoint being passed here. I'm thinking it'd be best to move this over into `path_login_test.go` where it's used.

**Comment by austingebauer** [comment]
I've added tests in 845d864. I also ran all other tests with success. See output below.

<details><summary>Verbose Test Output</summary>

```console
$ GOOGLE_CREDENTIALS=$(cat /Users/austingebauer/Workspace/hashico
rp/environments/vault/auth/gcp/vault-auth-tester.json) GOOGLE_CLOUD_PROJECT=hc-redacted go te
st -count=1 -v ./...
?       github.com/hashicorp/vault-plugin-auth-gcp      [no test files]
=== RUN   TestAuthorizeGCE
=== PAUSE TestAuthorizeGCE
=== RUN   TestGetIAMAlias
=== RUN   TestGetIAMAlias/invalid_type
=== RUN   TestGetIAMAlias/empty_type_goes_to_default
=== RUN   TestGetIAMAlias/default_type
=== RUN   TestGetIAMAlias/unique_id
--- PASS: TestGetIAMAlias (0.00s)
    --- PASS: TestGetIAMAlias/invalid_type (0.00s)
    --- PASS: TestGetIAMAlias/empty_type_goes_to_default (0.00s)
    --- PASS: TestGetIAMAlias/default_type (0.00s)
    --- PASS: TestGetIAMAlias/unique_id (0.00s)
=== RUN   TestGetGCEAlias
=== RUN   TestGetGCEAlias/invalid_type
=== RUN   TestGetGCEAlias/empty_type_goes_to_default
=== RUN   TestGetGCEAlias/default_type
=== RUN   TestGetGCEAlias/instance_id
--- PASS: TestGetGCEAlias (0.00s)
    --- PASS: TestGetGCEAlias/invalid_type (0.00s)
    --- PASS: TestGetGCEAlias/empty_type_goes_to_default (0.00s)
    --- PASS: TestGetGCEAlias/default_type (0.00s)
    --- PASS: TestGetGCEAlias/instance_id (0.00s)
=== RUN   TestZoneToRegion
=== PAUSE TestZoneToRegion
=== RUN   TestZoneFromSelfLink
=== PAUSE TestZoneFromSelfLink
=== RUN   TestBackend_PathConfigRead
=== RUN   TestBackend_PathConfigRead/field_validation
=== PAUSE TestBackend_PathConfigRead/field_validation
=== RUN   TestBackend_PathConfigRead/not_exist
=== PAUSE TestBackend_PathConfigRead/not_exist
=== RUN   TestBackend_PathConfigRead/exist
=== PAUSE TestBackend_PathConfigRead/exist
=== CONT  TestBackend_PathConfigRead/field_validation
=== CONT  TestBackend_PathConfigRead/exist
=== CONT  TestBackend_PathConfigRead/not_exist
--- PASS: TestBackend_PathConfigRead (0.00s)
    --- PASS: TestBackend_PathConfigRead/field_validation (0.00s)
    --- PASS: TestBackend_PathConfigRead/not_exist (0.00s)
    --- PASS: TestBackend_PathConfigRead/exist (0.00s)
=== RUN   TestBackend_PathConfigWrite
=== RUN   TestBackend_PathConfigWrite/field_validation
=== RUN   TestBackend_PathConfigWrite/not_exist
=== RUN   TestBackend_PathConfigWrite/bad_custom_endpoint
=== RUN   TestBackend_PathConfigWrite/exist
--- PASS: TestBackend_PathConfigWrite (0.00s)
    --- PASS: TestBackend_PathConfigWrite/field_validation (0.00s)
    --- PASS: TestBackend_PathConfigWrite/not_exist (0.00s)
    --- PASS: TestBackend_PathConfigWrite/bad_custom_endpoint (0.00s)
    --- PASS: TestBackend_PathConfigWrite/exist (0.00s)
=== RUN   TestConfig_Update
=== RUN   TestConfig_Update/empty
=== PAUSE TestConfig_Update/empty
=== RUN   TestConfig_Update/keeps_existing
=== PAUSE TestConfig_Update/keeps_existing
=== RUN   TestConfig_Update/overwrites_changes
=== PAUSE TestConfig_Update/overwrites_changes
=== RUN   TestConfig_Update/overwrites_and_new
=== PAUSE TestConfig_Update/overwrites_and_new
=== CONT  TestConfig_Update/empty
=== CONT  TestConfig_Update/keeps_existing
=== CONT  TestConfig_Update/overwrites_and_new
=== CONT  TestConfig_Update/overwrites_changes
--- PASS: TestConfig_Update (0.00s)
    --- PASS: TestConfig_Update/empty (0.00s)
    --- PASS: TestConfig_Update/keeps_existing (0.00s)
    --- PASS: TestConfig_Update/overwrites_and_new (0.00s)
    --- PASS: TestConfig_Update/overwrites_changes (0.00s)
=== RUN   TestLogin_IAM
=== PAUSE TestLogin_IAM
=== RUN   TestLogin_IAM_Custom_Endpoint
=== RUN   TestLogin_IAM_Custom_Endpoint/IAM_login_with_invalid_api_custom_endpoints_results_in_error
=== RUN   TestLogin_IAM_Custom_Endpoint/IAM_login_with_invalid_iam_custom_endpoints_results_in_error
=== RUN   TestLogin_IAM_Custom_Endpoint/IAM_login_with_invalid_crm_custom_endpoints_results_in_error
=== RUN   TestLogin_IAM_Custom_Endpoint/IAM_login_with_invalid_compute_custom_endpoints_results_in_success
=== RUN   TestLogin_IAM_Custom_Endpoint/IAM_login_with_default_custom_endpoints_results_in_success
=== RUN   TestLogin_IAM_Custom_Endpoint/IAM_login_with_empty_custom_endpoints_results_in_success
--- PASS: TestLogin_IAM_Custom_Endpoint (3.82s)
    --- PASS: TestLogin_IAM_Custom_Endpoint/IAM_login_with_invalid_api_custom_endpoints_results_in_error (0.42s)
    --- PASS: TestLogin_IAM_Custom_Endpoint/IAM_login_with_invalid_iam_custom_endpoints_results_in_error (0.34s)
    --- PASS: TestLogin_IAM_Custom_Endpoint/IAM_login_with_invalid_crm_custom_endpoints_results_in_error (0.47s)
    --- PASS: TestLogin_IAM_Custom_Endpoint/IAM_login_with_invalid_compute_custom_endpoints_results_in_success (0.92s)
    --- PASS: TestLogin_IAM_Custom_Endpoint/IAM_login_with_default_custom_endpoints_results_in_success (0.84s)
    --- PASS: TestLogin_IAM_Custom_Endpoint/IAM_login_with_empty_custom_endpoints_results_in_success (0.83s)
=== RUN   Test_Renew
--- PASS: Test_Renew (0.75s)
=== RUN   TestRoleUpdateIam
=== PAUSE TestRoleUpdateIam
=== RUN   TestRoleIam_Wildcard
=== PAUSE TestRoleIam_Wildcard
=== RUN   TestRoleIam_EditServiceAccounts
=== PAUSE TestRoleIam_EditServiceAccounts
=== RUN   TestRoleIam_MissingRequiredArgs
=== PAUSE TestRoleIam_MissingRequiredArgs
=== RUN   TestRoleIam_HasGceArgs
=== PAUSE TestRoleIam_HasGceArgs
=== RUN   TestRoleGce
=== PAUSE TestRoleGce
=== RUN   TestRoleGce_EditLabels
=== PAUSE TestRoleGce_EditLabels
=== RUN   TestRoleGce_DeprecatedFields
=== PAUSE TestRoleGce_DeprecatedFields
=== RUN   TestRole_MissingRequiredArgs
=== PAUSE TestRole_MissingRequiredArgs
=== RUN   TestRole_InvalidRoleType
--- PASS: TestRole_InvalidRoleType (0.00s)
=== RUN   TestRetrieveRole
=== RUN   TestRetrieveRole/TokenPeriod_upgrade
=== RUN   TestRetrieveRole/TokenPolicies_upgrade
=== RUN   TestRetrieveRole/bad_data
=== RUN   TestRetrieveRole/boundRegion_upgrade
=== RUN   TestRetrieveRole/boundZone_upgrade
=== RUN   TestRetrieveRole/boundInstanceGroup_upgrade
=== RUN   TestRetrieveRole/TTL_upgrade
=== RUN   TestRetrieveRole/not_found
=== RUN   TestRetrieveRole/storage_error
=== RUN   TestRetrieveRole/projectID_upgrade
=== RUN   TestRetrieveRole/MaxTTL_upgrade
=== RUN   TestRetrieveRole/storage_put_error
=== RUN   TestRetrieveRole/roleID_is_generated_when_one_does_not_exist
--- PASS: TestRetrieveRole (0.00s)
    --- PASS: TestRetrieveRole/TokenPeriod_upgrade (0.00s)
    --- PASS: TestRetrieveRole/TokenPolicies_upgrade (0.00s)
    --- PASS: TestRetrieveRole/bad_data (0.00s)
    --- PASS: TestRetrieveRole/boundRegion_upgrade (0.00s)
    --- PASS: TestRetrieveRole/boundZone_upgrade (0.00s)
    --- PASS: TestRetrieveRole/boundInstanceGroup_upgrade (0.00s)
    --- PASS: TestRetrieveRole/TTL_upgrade (0.00s)
    --- PASS: TestRetrieveRole/not_found (0.00s)
    --- PASS: TestRetrieveRole/storage_error (0.00s)
    --- PASS: TestRetrieveRole/projectID_upgrade (0.00s)
    --- PASS: TestRetrieveRole/MaxTTL_upgrade (0.00s)
    --- PASS: TestRetrieveRole/storage_put_error (0.00s)
    --- PASS: TestRetrieveRole/roleID_is_generated_when_one_does_not_exist (0.00s)
=== CONT  TestAuthorizeGCE
=== CONT  TestRoleIam_MissingRequiredArgs
=== CONT  TestRoleGce
=== CONT  TestZoneToRegion
=== RUN   TestZoneToRegion/0_us-central1-a_to_us-central1
=== PAUSE TestZoneToRegion/0_us-central1-a_to_us-central1
=== RUN   TestZoneToRegion/1_northamerica-northeast1-c_to_northamerica-northeast1
=== CONT  TestRoleIam_HasGceArgs
=== PAUSE TestZoneToRegion/1_northamerica-northeast1-c_to_northamerica-northeast1
=== RUN   TestZoneToRegion/2_europe-west3-c_to_europe-west3
=== PAUSE TestZoneToRegion/2_europe-west3-c_to_europe-west3
=== CONT  TestRole_MissingRequiredArgs
=== CONT  TestRoleGce_DeprecatedFields
=== RUN   TestAuthorizeGCE/labels_no_match_key
=== PAUSE TestAuthorizeGCE/labels_no_match_key
=== CONT  TestRoleIam_EditServiceAccounts
=== CONT  TestRoleIam_Wildcard
=== CONT  TestRoleUpdateIam
=== CONT  TestLogin_IAM
=== CONT  TestZoneFromSelfLink
=== CONT  TestRoleGce_EditLabels
=== RUN   TestZoneToRegion/3_us_err
=== RUN   TestRoleGce_DeprecatedFields/deprecated_fields_upgraded
=== RUN   TestAuthorizeGCE/labels_no_match_value
=== PAUSE TestAuthorizeGCE/labels_no_match_value
=== RUN   TestAuthorizeGCE/zone_as_self_link_exists
=== PAUSE TestAuthorizeGCE/zone_as_self_link_exists
=== RUN   TestAuthorizeGCE/zone_as_name_exists
=== PAUSE TestAuthorizeGCE/zone_as_name_exists
=== RUN   TestAuthorizeGCE/zone_as_self_link_no_exists
=== PAUSE TestAuthorizeGCE/zone_as_self_link_no_exists
=== RUN   TestAuthorizeGCE/zone_as_name_no_exists
=== PAUSE TestAuthorizeGCE/zone_as_name_no_exists
=== RUN   TestAuthorizeGCE/zone_as_invalid
=== PAUSE TestAuthorizeGCE/zone_as_invalid
=== RUN   TestAuthorizeGCE/region_as_self_link_exists
=== PAUSE TestAuthorizeGCE/region_as_self_link_exists
=== PAUSE TestZoneToRegion/3_us_err
--- PASS: TestRoleIam_MissingRequiredArgs (0.00s)
=== RUN   TestAuthorizeGCE/region_as_name_exists
=== PAUSE TestAuthorizeGCE/region_as_name_exists
=== RUN   TestAuthorizeGCE/region_as_self_link_no_exists
=== RUN   TestZoneToRegion/4__err
=== PAUSE TestAuthorizeGCE/region_as_self_link_no_exists
=== RUN   TestZoneFromSelfLink/0
=== PAUSE TestZoneToRegion/4__err
=== CONT  TestZoneToRegion/0_us-central1-a_to_us-central1
=== CONT  TestZoneToRegion/4__err
=== CONT  TestZoneToRegion/2_europe-west3-c_to_europe-west3
=== CONT  TestZoneToRegion/1_northamerica-northeast1-c_to_northamerica-northeast1
=== PAUSE TestRoleGce_DeprecatedFields/deprecated_fields_upgraded
=== RUN   TestRoleGce_DeprecatedFields/existing_storage_upgraded
=== PAUSE TestRoleGce_DeprecatedFields/existing_storage_upgraded
=== CONT  TestZoneToRegion/3_us_err
--- PASS: TestZoneToRegion (0.00s)
    --- PASS: TestZoneToRegion/0_us-central1-a_to_us-central1 (0.00s)
    --- PASS: TestZoneToRegion/4__err (0.00s)
    --- PASS: TestZoneToRegion/1_northamerica-northeast1-c_to_northamerica-northeast1 (0.00s)
    --- PASS: TestZoneToRegion/2_europe-west3-c_to_europe-west3 (0.00s)
    --- PASS: TestZoneToRegion/3_us_err (0.00s)
=== PAUSE TestZoneFromSelfLink/0
=== RUN   TestZoneFromSelfLink/1
=== PAUSE TestZoneFromSelfLink/1
=== RUN   TestAuthorizeGCE/region_as_name_no_exists
=== RUN   TestZoneFromSelfLink/2
=== CONT  TestRoleGce_DeprecatedFields/deprecated_fields_upgraded
=== PAUSE TestZoneFromSelfLink/2
=== CONT  TestZoneFromSelfLink/0
=== CONT  TestZoneFromSelfLink/1
--- PASS: TestRole_MissingRequiredArgs (0.00s)
=== PAUSE TestAuthorizeGCE/region_as_name_no_exists
=== CONT  TestZoneFromSelfLink/2
=== CONT  TestRoleGce_DeprecatedFields/existing_storage_upgraded
--- PASS: TestRoleIam_HasGceArgs (0.00s)
=== RUN   TestAuthorizeGCE/region_as_invalid
=== PAUSE TestAuthorizeGCE/region_as_invalid
--- PASS: TestRoleIam_Wildcard (0.00s)
--- PASS: TestRoleGce (0.00s)
--- PASS: TestZoneFromSelfLink (0.00s)
    --- PASS: TestZoneFromSelfLink/0 (0.00s)
    --- PASS: TestZoneFromSelfLink/1 (0.00s)
    --- PASS: TestZoneFromSelfLink/2 (0.00s)
=== RUN   TestAuthorizeGCE/bound_instance_groups_unbound
=== PAUSE TestAuthorizeGCE/bound_instance_groups_unbound
=== RUN   TestAuthorizeGCE/bound_instance_groups_empty_bound_zones
--- PASS: TestRoleIam_EditServiceAccounts (0.00s)
--- PASS: TestRoleUpdateIam (0.00s)
=== PAUSE TestAuthorizeGCE/bound_instance_groups_empty_bound_zones
=== RUN   TestAuthorizeGCE/bound_instance_groups_no_exist_bound_zones
--- PASS: TestRoleGce_DeprecatedFields (0.00s)
    --- PASS: TestRoleGce_DeprecatedFields/existing_storage_upgraded (0.00s)
    --- PASS: TestRoleGce_DeprecatedFields/deprecated_fields_upgraded (0.00s)
=== PAUSE TestAuthorizeGCE/bound_instance_groups_no_exist_bound_zones
=== RUN   TestAuthorizeGCE/bound_instance_groups_empty_bound_regions
=== PAUSE TestAuthorizeGCE/bound_instance_groups_empty_bound_regions
=== RUN   TestAuthorizeGCE/bound_instance_groups_no_exist_bound_regions
=== PAUSE TestAuthorizeGCE/bound_instance_groups_no_exist_bound_regions
=== RUN   TestAuthorizeGCE/bound_instance_groups_no_contains_instance
=== PAUSE TestAuthorizeGCE/bound_instance_groups_no_contains_instance
=== RUN   TestAuthorizeGCE/bound_service_account_no_exist
=== PAUSE TestAuthorizeGCE/bound_service_account_no_exist
=== RUN   TestAuthorizeGCE/bound_service_account_id
=== PAUSE TestAuthorizeGCE/bound_service_account_id
=== RUN   TestAuthorizeGCE/bound_service_account_email
=== PAUSE TestAuthorizeGCE/bound_service_account_email
=== RUN   TestAuthorizeGCE/success_zone_binding
=== PAUSE TestAuthorizeGCE/success_zone_binding
=== RUN   TestAuthorizeGCE/success_region_binding
=== PAUSE TestAuthorizeGCE/success_region_binding
=== RUN   TestAuthorizeGCE/success_instance_group_zone_binding
=== PAUSE TestAuthorizeGCE/success_instance_group_zone_binding
--- PASS: TestRoleGce_EditLabels (0.00s)
=== RUN   TestAuthorizeGCE/success_instance_group_region_binding
=== PAUSE TestAuthorizeGCE/success_instance_group_region_binding
=== CONT  TestAuthorizeGCE/labels_no_match_key
=== CONT  TestAuthorizeGCE/success_instance_group_zone_binding
=== CONT  TestAuthorizeGCE/region_as_name_no_exists
=== CONT  TestAuthorizeGCE/region_as_self_link_no_exists
=== CONT  TestAuthorizeGCE/bound_service_account_email
=== CONT  TestAuthorizeGCE/success_region_binding
=== CONT  TestAuthorizeGCE/region_as_invalid
=== CONT  TestAuthorizeGCE/zone_as_name_no_exists
=== CONT  TestAuthorizeGCE/zone_as_invalid
=== CONT  TestAuthorizeGCE/success_instance_group_region_binding
=== CONT  TestAuthorizeGCE/zone_as_self_link_exists
=== CONT  TestAuthorizeGCE/zone_as_name_exists
=== CONT  TestAuthorizeGCE/success_zone_binding
=== CONT  TestAuthorizeGCE/labels_no_match_value
=== CONT  TestAuthorizeGCE/bound_service_account_no_exist
=== CONT  TestAuthorizeGCE/bound_instance_groups_no_contains_instance
=== CONT  TestAuthorizeGCE/bound_instance_groups_no_exist_bound_regions
=== CONT  TestAuthorizeGCE/bound_instance_groups_empty_bound_regions
=== CONT  TestAuthorizeGCE/bound_instance_groups_no_exist_bound_zones
=== CONT  TestAuthorizeGCE/bound_instance_groups_empty_bound_zones
=== CONT  TestAuthorizeGCE/bound_instance_groups_unbound
=== CONT  TestAuthorizeGCE/zone_as_self_link_no_exists
=== CONT  TestAuthorizeGCE/region_as_name_exists
=== CONT  TestAuthorizeGCE/region_as_self_link_exists
=== CONT  TestAuthorizeGCE/bound_service_account_id
--- PASS: TestAuthorizeGCE (0.00s)
    --- PASS: TestAuthorizeGCE/labels_no_match_key (0.00s)
    --- PASS: TestAuthorizeGCE/success_instance_group_zone_binding (0.00s)
    --- PASS: TestAuthorizeGCE/bound_service_account_email (0.00s)
    --- PASS: TestAuthorizeGCE/region_as_name_no_exists (0.00s)
    --- PASS: TestAuthorizeGCE/region_as_self_link_no_exists (0.00s)
    --- PASS: TestAuthorizeGCE/region_as_invalid (0.00s)
    --- PASS: TestAuthorizeGCE/success_region_binding (0.00s)
    --- PASS: TestAuthorizeGCE/zone_as_invalid (0.00s)
    --- PASS: TestAuthorizeGCE/zone_as_name_no_exists (0.00s)
    --- PASS: TestAuthorizeGCE/zone_as_self_link_exists (0.00s)
    --- PASS: TestAuthorizeGCE/zone_as_name_exists (0.00s)
    --- PASS: TestAuthorizeGCE/success_instance_group_region_binding (0.00s)
    --- PASS: TestAuthorizeGCE/success_zone_binding (0.00s)
    --- PASS: TestAuthorizeGCE/labels_no_match_value (0.00s)
    --- PASS: TestAuthorizeGCE/bound_instance_groups_no_contains_instance (0.00s)
    --- PASS: TestAuthorizeGCE/bound_service_account_no_exist (0.00s)
    --- PASS: TestAuthorizeGCE/bound_instance_groups_no_exist_bound_regions (0.00s)
    --- PASS: TestAuthorizeGCE/bound_instance_groups_empty_bound_regions (0.00s)
    --- PASS: TestAuthorizeGCE/bound_instance_groups_no_exist_bound_zones (0.00s)
    --- PASS: TestAuthorizeGCE/bound_instance_groups_empty_bound_zones (0.00s)
    --- PASS: TestAuthorizeGCE/bound_instance_groups_unbound (0.00s)
    --- PASS: TestAuthorizeGCE/zone_as_self_link_no_exists (0.00s)
    --- PASS: TestAuthorizeGCE/region_as_name_exists (0.00s)
    --- PASS: TestAuthorizeGCE/region_as_self_link_exists (0.00s)
    --- PASS: TestAuthorizeGCE/bound_service_account_id (0.00s)
=== RUN   TestLogin_IAM/not_bound
=== PAUSE TestLogin_IAM/not_bound
=== RUN   TestLogin_IAM/not_bound_project
=== PAUSE TestLogin_IAM/not_bound_project
=== RUN   TestLogin_IAM/no_policies
=== PAUSE TestLogin_IAM/no_policies
=== RUN   TestLogin_IAM/expire_late
=== PAUSE TestLogin_IAM/expire_late
=== RUN   TestLogin_IAM/group_aliases
=== PAUSE TestLogin_IAM/group_aliases
=== RUN   TestLogin_IAM/wildcard
=== PAUSE TestLogin_IAM/wildcard
=== RUN   TestLogin_IAM/ttl
=== PAUSE TestLogin_IAM/ttl
=== RUN   TestLogin_IAM/max_ttl
=== PAUSE TestLogin_IAM/max_ttl
=== RUN   TestLogin_IAM/period
=== PAUSE TestLogin_IAM/period
=== RUN   TestLogin_IAM/jwt_already_expired
=== PAUSE TestLogin_IAM/jwt_already_expired
=== CONT  TestLogin_IAM/not_bound
=== CONT  TestLogin_IAM/wildcard
=== CONT  TestLogin_IAM/jwt_already_expired
=== CONT  TestLogin_IAM/max_ttl
=== CONT  TestLogin_IAM/period
=== CONT  TestLogin_IAM/expire_late
=== CONT  TestLogin_IAM/group_aliases
=== CONT  TestLogin_IAM/ttl
=== CONT  TestLogin_IAM/not_bound_project
=== CONT  TestLogin_IAM/no_policies
--- PASS: TestLogin_IAM (0.47s)
    --- PASS: TestLogin_IAM/jwt_already_expired (0.13s)
    --- PASS: TestLogin_IAM/expire_late (0.25s)
    --- PASS: TestLogin_IAM/max_ttl (0.40s)
    --- PASS: TestLogin_IAM/no_policies (0.44s)
    --- PASS: TestLogin_IAM/period (0.45s)
    --- PASS: TestLogin_IAM/wildcard (0.45s)
    --- PASS: TestLogin_IAM/ttl (0.49s)
    --- PASS: TestLogin_IAM/not_bound_project (0.50s)
    --- PASS: TestLogin_IAM/not_bound (0.51s)
    --- PASS: TestLogin_IAM/group_aliases (0.87s)
PASS
ok      github.com/hashicorp/vault-plugin-auth-gcp/plugin       6.396s
?       github.com/hashicorp/vault-plugin-auth-gcp/plugin/cache [no test files]
```

</details>

File: plugin/backend.go (hunk 3 above)
Comment (austingebauer) [reply plugin/backend.go]: I moved this to `path_login_test.go` in 845d864. I also moved some other code there that was only being used in tests.

File: plugin/gcp_config.go (hunk 2 above)
Comment (fairclothjm) [reply plugin/gcp_config.go:96]: _If_ we did validation maybe just verify it is a valid URL? But I might lean more towards letting the user set what they want and allow the API call itself to report any errors.

**Review by robmonte — DISMISSED** [review]
Everything looks good to me, just some minor inquiries as nothing else stood out.

File: README.md [hunk 4]
```diff
@@ -126,8 +126,9 @@ permissions listed below. The following steps assume you have
     ```text
     $ export GOOGLE_CLOUD_PROJECT=my-project # replace with your project ID
     ```
+The remaining steps may also be executed by running `. scripts/cred_setup.sh`.
```
[the remark this answers is not in the document]
Comment (fairclothjm) [reply README.md:129]: `.` is POSIX compliant and not all shells have `source`. I don't know if we are strict about that though.

**Review by fairclothjm — APPROVED** [review]
LGTM! I created a google project and ran the tests -- All passed

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (robmonte) [reply README.md:129]: Oh that's good to learn, I incorrectly thought it was the opposite!

[the remark this answers is not in the document]
Comment (austingebauer) [reply README.md:129]: I'll likely leave it with `.` since it's more portable (TIL as well). Good consideration though, @robmonte.

**Review by fairclothjm — APPROVED** [review]

File: plugin/gcp_config.go (hunk 2 above)
Comment (austingebauer) [reply plugin/gcp_config.go:96]: I tested out some bogus `option.WithEndpoint()` values, and Google's client doesn't validate that it's value has some specific format. So I'm +1 on letting the user set what they want here. The error will be apparent from the result of the API call. Thanks!

**Pull request merged**